spec/storage/redis_spec.rb

```diff
@@ -46,6 +47,14 @@
     end
   end
 
+  context 'when use with Redis Cluster' do
+    let(:client) { Redis::Cluster.new(cluster: [{ host: '127.0.0.1', port: '7000' }]) }
```

Comment:
@justinhoward Have you written a test involves `Redis::Cluster` before? This works but it really tries to connect to a Redis cluster node:

<img width="1234" alt="CleanShot 2022-02-19 at 15 07 44@2x" src="https://user-images.githubusercontent.com/1000669/154788757-79678ce9-d5bb-4ecc-9d20-bd76c474d4e2.png">

I also explored doing a `FakeRakeCluster.new` but there are too many methods to stub and gave up 🤔

Reply:
Redis clusters have a [superset of commands](https://redis.io/commands/cluster-nodes). Unless we use a real Redis cluster to test (CI does not), cluster commands will fail.

However, I don't think we need to even reach the point where we would connect to Redis. I see no reason to allow execution to continue at all if a cluster is given. It's a bit different than the warnings currently printed to stderr. In this case, we're detecting a completely incompatible configuration, so we should raise an error instead of printing a warning.

Reply:
My side project is using Redis cluster with Faulty right now and it is working (although I disable checking redis and pool options because I encountered this crash), so I think we should still print a warning instead of raising an error?

Reply:
When I run the tests on a Redis cluster, I get `CROSSLOT` and `MOVED` errors as you would expect when manipulating multiple keys in a pipeline. Here's a small sample

```
  1) circuits with redis storage behaves like circuit runs a circuit with no errors
     Got 0 failures and 2 other errors:
     Shared Example Group: "circuit" called from ./spec/circuit_spec.rb:356

     1.1) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            MOVED 15157 127.0.0.1:6379
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:162:in `call'
          # ./lib/faulty/patch/redis.rb:59:in `block in call'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:59:in `call'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:263:in `block in send_command'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:262:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:262:in `send_command'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/commands/strings.rb:97:in `set'
          # ./lib/faulty/storage/redis.rb:113:in `block in set_options'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:112:in `set_options'
          # ./lib/faulty/circuit.rb:403:in `push_options'
          # ./lib/faulty/circuit.rb:317:in `run'
          # ./spec/circuit_spec.rb:50:in `block (3 levels) in <top (required)>'

     1.2) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            CROSSSLOT Keys in request don't hash to the same slot
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:246:in `call_pipelined'
          # ./lib/faulty/patch/redis.rb:69:in `block in call_pipelined'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:69:in `call_pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:200:in `block in call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:344:in `with_reconnect'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:198:in `call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:170:in `block in pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `block in synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:164:in `pipelined'
          # ./lib/faulty/storage/redis.rb:389:in `block in pipe'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:389:in `pipe'
          # ./lib/faulty/storage/redis.rb:208:in `reset'
          # ./lib/faulty/circuit.rb:360:in `reset!'
          # ./spec/circuit_spec.rb:354:in `block (3 levels) in <top (required)>'

  2) circuits with redis storage behaves like circuit gets an ok result with try_run
     Got 0 failures and 2 other errors:
     Shared Example Group: "circuit" called from ./spec/circuit_spec.rb:356

     2.1) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            MOVED 15157 127.0.0.1:6379
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:162:in `call'
          # ./lib/faulty/patch/redis.rb:59:in `block in call'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:59:in `call'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:263:in `block in send_command'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:262:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:262:in `send_command'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/commands/strings.rb:97:in `set'
          # ./lib/faulty/storage/redis.rb:113:in `block in set_options'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:112:in `set_options'
          # ./lib/faulty/circuit.rb:403:in `push_options'
          # ./lib/faulty/circuit.rb:317:in `run'
          # ./lib/faulty/circuit.rb:282:in `try_run'
          # ./spec/circuit_spec.rb:54:in `block (3 levels) in <top (required)>'

     2.2) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            CROSSSLOT Keys in request don't hash to the same slot
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:246:in `call_pipelined'
          # ./lib/faulty/patch/redis.rb:69:in `block in call_pipelined'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:69:in `call_pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:200:in `block in call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:344:in `with_reconnect'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:198:in `call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:170:in `block in pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `block in synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:164:in `pipelined'
          # ./lib/faulty/storage/redis.rb:389:in `block in pipe'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:389:in `pipe'
          # ./lib/faulty/storage/redis.rb:208:in `reset'
          # ./lib/faulty/circuit.rb:360:in `reset!'
          # ./spec/circuit_spec.rb:354:in `block (3 levels) in <top (required)>'

  3) circuits with redis storage behaves like circuit captures an error with try_run
     Got 0 failures and 2 other errors:
     Shared Example Group: "circuit" called from ./spec/circuit_spec.rb:356

     3.1) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            MOVED 15157 127.0.0.1:6379
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:162:in `call'
          # ./lib/faulty/patch/redis.rb:59:in `block in call'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:59:in `call'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:263:in `block in send_command'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:262:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:262:in `send_command'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/commands/strings.rb:97:in `set'
          # ./lib/faulty/storage/redis.rb:113:in `block in set_options'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:112:in `set_options'
          # ./lib/faulty/circuit.rb:403:in `push_options'
          # ./lib/faulty/circuit.rb:317:in `run'
          # ./lib/faulty/circuit.rb:282:in `try_run'
          # ./spec/circuit_spec.rb:60:in `block (3 levels) in <top (required)>'

     3.2) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            CROSSSLOT Keys in request don't hash to the same slot
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:246:in `call_pipelined'
          # ./lib/faulty/patch/redis.rb:69:in `block in call_pipelined'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:69:in `call_pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:200:in `block in call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:344:in `with_reconnect'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:198:in `call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:170:in `block in pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `block in synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:164:in `pipelined'
          # ./lib/faulty/storage/redis.rb:389:in `block in pipe'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:389:in `pipe'
          # ./lib/faulty/storage/redis.rb:208:in `reset'
          # ./lib/faulty/circuit.rb:360:in `reset!'
          # ./spec/circuit_spec.rb:354:in `block (3 levels) in <top (required)>'

  4) circuits with redis storage behaves like circuit raises a CircuitFailureError when an error is raised
     Got 0 failures and 2 other errors:
     Shared Example Group: "circuit" called from ./spec/circuit_spec.rb:356

     4.1) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            MOVED 786 127.0.0.1:6379
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:246:in `call_pipelined'
          # ./lib/faulty/patch/redis.rb:69:in `block in call_pipelined'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:69:in `call_pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:200:in `block in call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:344:in `with_reconnect'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:198:in `call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:170:in `block in pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `block in synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:164:in `pipelined'
          # ./lib/faulty/storage/redis.rb:389:in `block in pipe'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:389:in `pipe'
          # ./lib/faulty/storage/redis.rb:226:in `status'
          # ./lib/faulty/circuit.rb:372:in `status'
          # ./lib/faulty/circuit.rb:187:in `inspect'
          # ./spec/circuit_spec.rb:66:in `block (3 levels) in <top (required)>'

     4.2) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            CROSSSLOT Keys in request don't hash to the same slot
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:246:in `call_pipelined'
          # ./lib/faulty/patch/redis.rb:69:in `block in call_pipelined'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:69:in `call_pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:200:in `block in call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:344:in `with_reconnect'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:198:in `call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:170:in `block in pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `block in synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:164:in `pipelined'
          # ./lib/faulty/storage/redis.rb:389:in `block in pipe'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:389:in `pipe'
          # ./lib/faulty/storage/redis.rb:208:in `reset'
          # ./lib/faulty/circuit.rb:360:in `reset!'
          # ./spec/circuit_spec.rb:354:in `block (3 levels) in <top (required)>'

  5) circuits with redis storage behaves like circuit raises a CircuitTrippedError when the threshold is passed
     Got 0 failures and 2 other errors:
     Shared Example Group: "circuit" called from ./spec/circuit_spec.rb:356

     5.1) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            MOVED 786 127.0.0.1:6379
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:246:in `call_pipelined'
          # ./lib/faulty/patch/redis.rb:69:in `block in call_pipelined'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:69:in `call_pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:200:in `block in call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:344:in `with_reconnect'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:198:in `call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:170:in `block in pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `block in synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:164:in `pipelined'
          # ./lib/faulty/storage/redis.rb:389:in `block in pipe'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:389:in `pipe'
          # ./lib/faulty/storage/redis.rb:226:in `status'
          # ./lib/faulty/circuit.rb:372:in `status'
          # ./lib/faulty/circuit.rb:187:in `inspect'
          # ./spec/circuit_spec.rb:76:in `block (3 levels) in <top (required)>'

     5.2) Failure/Error: faulty_run { super }
          
          Redis::CommandError:
            CROSSSLOT Keys in request don't hash to the same slot
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:246:in `call_pipelined'
          # ./lib/faulty/patch/redis.rb:69:in `block in call_pipelined'
          # ./lib/faulty/patch/base.rb:36:in `faulty_run'
          # ./lib/faulty/patch/redis.rb:69:in `call_pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:200:in `block in call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:344:in `with_reconnect'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis/client.rb:198:in `call_pipeline'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:170:in `block in pipelined'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `block in synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:258:in `synchronize'
          # ./vendor/bundle/ruby/3.0.0/gems/redis-4.6.0/lib/redis.rb:164:in `pipelined'
          # ./lib/faulty/storage/redis.rb:389:in `block in pipe'
          # ./lib/faulty/storage/redis.rb:380:in `redis'
          # ./lib/faulty/storage/redis.rb:389:in `pipe'
          # ./lib/faulty/storage/redis.rb:208:in `reset'
          # ./lib/faulty/circuit.rb:360:in `reset!'
          # ./spec/circuit_spec.rb:354:in `block (3 levels) in <top (required)>'
```